Mathlib/CategoryTheory/Sites/Canonical.lean

```diff
@@ -0,0 +1,263 @@
+axiom) forms the bulk of this file.
+
+This generalises to a set of presheaves, giving the topology `finestTopology Ps` which is the
+finest topology for which every presheaf in `Ps` is a sheaf.
+Using `Ps` as the set of representable presheaves defines the `canonicalTopology`: the finest
+topology for which every representable is a sheaf.
+
+A Grothendieck topology is called `Subcanonical` if it is smaller than the canonical topology,
+equivalently it is subcanonical iff every representable presheaf is a sheaf.
+
+## References
+* https://ncatlab.org/nlab/show/canonical+topology
+* https://ncatlab.org/nlab/show/subcanonical+coverage
+* https://stacks.math.columbia.edu/tag/00Z9
+* https://math.stackexchange.com/a/358709/
+-/
+
+
+universe v u
+
+namespace CategoryTheory
+
+open CategoryTheory Category Limits Sieve Classical
+
+variable {C : Type u} [Category.{v} C]
+
+namespace Sheaf
+
+variable {P : Cᵒᵖ ⥤ Type v}
+
+variable {X Y : C} {S : Sieve X} {R : Presieve X}
+
+variable (J J₂ : GrothendieckTopology C)
+
+/--
+To show `P` is a sheaf for the binding of `U` with `B`, it suffices to show that `P` is a sheaf for
+`U`, that `P` is a sheaf for each sieve in `B`, and that it is separated for any pullback of any
+sieve in `B`.
+
+This is mostly an auxiliary lemma to show `isSheafFor_trans`.
+Adapted from [Elephant], Lemma C2.1.7(i) with suggestions as mentioned in
+https://math.stackexchange.com/a/358709/
+-/
+theorem isSheafFor_bind (P : Cᵒᵖ ⥤ Type v) (U : Sieve X) (B : ∀ ⦃Y⦄ ⦃f : Y ⟶ X⦄, U f → Sieve Y)
+    (hU : Presieve.IsSheafFor P (U : Presieve X))
+    (hB : ∀ ⦃Y⦄ ⦃f : Y ⟶ X⦄ (hf : U f), Presieve.IsSheafFor P (B hf : Presieve Y))
+    (hB' : ∀ ⦃Y⦄ ⦃f : Y ⟶ X⦄ (h : U f) ⦃Z⦄ (g : Z ⟶ Y),
+      Presieve.IsSeparatedFor P (((B h).pullback g) : Presieve Z)) :
+    Presieve.IsSheafFor P (Sieve.bind (U : Presieve X) B : Presieve X) := by
+  intro s hs
+  let y : ∀ ⦃Y⦄ ⦃f : Y ⟶ X⦄ (hf : U f), Presieve.FamilyOfElements P (B hf : Presieve Y) :=
+    fun Y f hf Z g hg => s _ (Presieve.bind_comp _ _ hg)
+  have hy : ∀ ⦃Y⦄ ⦃f : Y ⟶ X⦄ (hf : U f), (y hf).Compatible := by
+    intro Y f H Y₁ Y₂ Z g₁ g₂ f₁ f₂ hf₁ hf₂ comm
+    apply hs
+    apply reassoc_of% comm
+  let t : Presieve.FamilyOfElements P (U : Presieve X) :=
+    fun Y f hf => (hB hf).amalgamate (y hf) (hy hf)
+  have ht : ∀ ⦃Y⦄ ⦃f : Y ⟶ X⦄ (hf : U f), (y hf).IsAmalgamation (t f hf) := fun Y f hf =>
+    (hB hf).isAmalgamation _
+  have hT : t.Compatible := by
+    rw [Presieve.compatible_iff_sieveCompatible]
+    intro Z W f h hf
+    apply (hB (U.downward_closed hf h)).isSeparatedFor.ext
+    intro Y l hl
+    apply (hB' hf (l ≫ h)).ext
+    intro M m hm
+    have : bind U B (m ≫ l ≫ h ≫ f) := by
+      -- porting note: had to make explicit the parameter `((m ≫ l ≫ h) ≫ f)` and
+      -- using `by exact`
+      have : bind U B ((m ≫ l ≫ h) ≫ f) := by exact Presieve.bind_comp f hf hm
+      -- porting note: was `simpa using this`, see also following porting notes
```

Comment:
This seems to work for me:
```
have : bind U B ((m ≫ l ≫ h) ≫ f) := by exact Presieve.bind_comp f hf hm
simpa using this
```

Reply:
By the way, for readability you can add the command
```
pp_extended_field_notation Sieve.arrows
```
to get `arrows` to show up with dot notation. It should go after the `structure` declaration.

Reply:
Thanks for these remarks!